File: src/com/goodtrendltd/HolySongs/HTML5WebView.java
package com.goodtrendltd.HolySongs;

import android.app.Activity;
import android.app.ProgressDialog;
import android.content.Context;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.net.Uri;
import android.net.http.SslError;
import android.os.Bundle;
import android.util.AttributeSet;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.Window;
import android.webkit.GeolocationPermissions;
import android.webkit.SslErrorHandler;
import android.webkit.WebChromeClient;
import android.webkit.WebResourceError;
import android.webkit.WebResourceRequest;
import android.webkit.WebSettings;
import android.webkit.WebView;
import android.webkit.WebViewClient;
import android.widget.FrameLayout;

/** WebView host used by the retained Java video-search screen. */
public class HTML5WebView extends WebView {

    private Context mContext;
    private MyWebChromeClient mWebChromeClient;
    private MyWebViewClient mWebViewClient;
    private View mCustomView;
    private FrameLayout mCustomViewContainer;
    private WebChromeClient.CustomViewCallback mCustomViewCallback;

    private FrameLayout mContentView;
    private FrameLayout mBrowserFrameLayout;
    private FrameLayout mLayout;

    private ProgressDialog pd;
    private String loadingUrl;
    private boolean loading;
    private boolean hostActive = true;
    private boolean released;
    private boolean destroyed;
    private boolean handlingProgressCancel;

    static final String LOGTAG = "HTML5WebView";

    private void init(Context context) {
        mContext = context;
        mLayout = new FrameLayout(context);

        mBrowserFrameLayout = (FrameLayout) LayoutInflater.from(context).inflate(R.layout.video_html5_screen, null);
        mContentView = (FrameLayout) mBrowserFrameLayout.findViewById(R.id.main_content);
        mCustomViewContainer = (FrameLayout) mBrowserFrameLayout.findViewById(R.id.fullscreen_custom_content);

        mLayout.addView(mBrowserFrameLayout, COVER_SCREEN_PARAMS);

        mWebChromeClient = new MyWebChromeClient();
        setWebChromeClient(mWebChromeClient);

        mWebViewClient = new MyWebViewClient();
        setWebViewClient(mWebViewClient);

        WebSettings s = getSettings();
        s.setLayoutAlgorithm(WebSettings.LayoutAlgorithm.NARROW_COLUMNS);
        // Password saving is not used by video search and is deprecated on current WebView.
        s.setSaveFormData(true);
        s.setJavaScriptEnabled(true);
        s.setDomStorageEnabled(true);

        mContentView.addView(this);
    }

    public HTML5WebView(Context context) {
        super(context);
        init(context);
    }

    public HTML5WebView(Context context, AttributeSet attrs) {
        super(context, attrs);
        init(context);
    }

    public HTML5WebView(Context context, AttributeSet attrs, int defStyle) {
        super(context, attrs, defStyle);
        init(context);
    }

    public FrameLayout getLayout() {
        return mLayout;
    }

    public boolean inCustomView() {
        return mCustomView != null;
    }

    public boolean isLoading() {
        return loading;
    }

    boolean isReleased() {
        return released;
    }

    WebViewClient getWebViewClientForTest() {
        return mWebViewClient;
    }

    WebChromeClient getWebChromeClientForTest() {
        return mWebChromeClient;
    }

    void setHostActive(boolean active) {
        if (released) {
            return;
        }
        hostActive = active;
        if (!active) {
            dismissProgress();
        }
    }

    void cancelLoading() {
        if (released) {
            return;
        }
        stopLoading();
        settleLoading();
    }

    // Test-only seam exercises the same ProgressDialog cancellation callback used by users.
    void cancelProgressDialogForTest() {
        if (pd != null) {
            pd.cancel();
        } else {
            cancelLoading();
        }
    }

    private void cancelLoadingFromProgressDialog() {
        if (released || handlingProgressCancel) {
            return;
        }
        handlingProgressCancel = true;
        try {
            cancelLoading();
        } finally {
            handlingProgressCancel = false;
        }
    }

    void pauseForHost() {
        if (!released) {
            onPause();
        }
    }

    void resumeForHost() {
        if (!released) {
            onResume();
        }
    }

    public void hideCustomView() {
        if (!released) {
            mWebChromeClient.onHideCustomView();
        }
    }

    /** Idempotent final release; temporary Activity pauses must not call this. */
    void release() {
        if (released) {
            return;
        }
        released = true;
        hostActive = false;
        settleLoading();
        hideCustomViewForRelease();
        if (mContentView != null) {
            mContentView.removeView(this);
        }
        if (getParent() instanceof ViewGroup) {
            ((ViewGroup) getParent()).removeView(this);
        }
        if (!destroyed) {
            destroyed = true;
            super.destroy();
        }
        mContext = null;
        mContentView = null;
        mBrowserFrameLayout = null;
        mCustomViewContainer = null;
        mLayout = null;
    }

    private void hideCustomViewForRelease() {
        if (mCustomView == null) {
            return;
        }
        mCustomView.setVisibility(View.GONE);
        if (mCustomViewContainer != null) {
            mCustomViewContainer.removeView(mCustomView);
            mCustomViewContainer.setVisibility(View.GONE);
        }
        WebChromeClient.CustomViewCallback callback = mCustomViewCallback;
        mCustomView = null;
        mCustomViewCallback = null;
        if (callback != null) {
            callback.onCustomViewHidden();
        }
    }

    private void settleLoading() {
        loading = false;
        loadingUrl = null;
        dismissProgress();
    }

    private void dismissProgress() {
        ProgressDialog dialog = pd;
        pd = null;
        if (dialog != null && dialog.isShowing()) {
            try {
                dialog.dismiss();
            } catch (RuntimeException ignored) {
                // The host window may already have gone away.
            }
        }
    }

    private boolean isLiveForUi() {
        return !released && hostActive && mContext instanceof Activity && !((Activity) mContext).isFinishing();
    }

    private class MyWebChromeClient extends WebChromeClient {
        private Bitmap mDefaultVideoPoster;
        private View mVideoProgressView;

        @Override
        public void onShowCustomView(View view, WebChromeClient.CustomViewCallback callback) {
            if (released || !hostActive || view == null) {
                if (callback != null) {
                    callback.onCustomViewHidden();
                }
                return;
            }
            HTML5WebView.this.setVisibility(View.GONE);
            if (mCustomView != null) {
                if (callback != null) {
                    callback.onCustomViewHidden();
                }
                return;
            }
            mCustomViewContainer.addView(view);
            mCustomView = view;
            mCustomViewCallback = callback;
            mCustomViewContainer.setVisibility(View.VISIBLE);
        }

        @Override
        public void onHideCustomView() {
            if (mCustomView == null) {
                return;
            }
            mCustomView.setVisibility(View.GONE);
            if (mCustomViewContainer != null) {
                mCustomViewContainer.removeView(mCustomView);
                mCustomViewContainer.setVisibility(View.GONE);
            }
            WebChromeClient.CustomViewCallback callback = mCustomViewCallback;
            mCustomView = null;
            mCustomViewCallback = null;
            if (callback != null) {
                callback.onCustomViewHidden();
            }
            if (!released) {
                HTML5WebView.this.setVisibility(View.VISIBLE);
            }
        }

        @Override
        public Bitmap getDefaultVideoPoster() {
            if (mDefaultVideoPoster == null && !released) {
                mDefaultVideoPoster = BitmapFactory.decodeResource(getResources(), R.drawable.default_video_poster);
            }
            return mDefaultVideoPoster;
        }

        @Override
        public View getVideoLoadingProgressView() {
            if (mVideoProgressView == null && !released) {
                mVideoProgressView = LayoutInflater.from(mContext).inflate(R.layout.video_loading_progress, null);
            }
            return mVideoProgressView;
        }

        @Override
        public void onReceivedTitle(WebView view, String title) {
            if (isLiveForUi()) {
                ((Activity) mContext).setTitle(title);
            }
        }

        @Override
        public void onProgressChanged(WebView view, int newProgress) {
            if (isLiveForUi()) {
                ((Activity) mContext).getWindow().setFeatureInt(Window.FEATURE_PROGRESS, newProgress * 100);
            }
        }

        @Override
        public void onGeolocationPermissionsShowPrompt(String origin, GeolocationPermissions.Callback callback) {
            if (callback != null) {
                callback.invoke(origin, false, false);
            }
        }
    }

    private class MyWebViewClient extends WebViewClient {
        @Override
        public boolean shouldOverrideUrlLoading(WebView view, String url) {
            if (!isHttpOrHttpsUrl(url)) {
                // Consume custom/file/content/about schemes: never hand them to an external
                // activity or an implicit browser. HTTP(S) remains in this WebView.
                return true;
            }
            if (!released) {
                Log.i(LOGTAG, "shouldOverrideUrlLoading: " + url);
            }
            return false;
        }

        @Override
        public boolean shouldOverrideUrlLoading(WebView view, WebResourceRequest request) {
            String url = request == null || request.getUrl() == null
                    ? null : request.getUrl().toString();
            if (!isHttpOrHttpsUrl(url)) {
                return true;
            }
            if (!released) {
                Log.i(LOGTAG, "shouldOverrideUrlLoading: " + url);
            }
            return false;
        }

        @Override
        public void onPageStarted(WebView view, String url, Bitmap favicon) {
            if (released) {
                return;
            }
            loading = true;
            loadingUrl = url;
            if (isLiveForUi()) {
                dismissProgress();
                try {
                    pd = ProgressDialog.show(mContext, "", mContext.getString(R.string.loading_video), true);
                    pd.setCancelable(true);
                    pd.setOnCancelListener(dialog -> cancelLoadingFromProgressDialog());
                } catch (RuntimeException ignored) {
                    pd = null;
                }
            }
        }

        @Override
        public void onPageFinished(WebView view, String url) {
            if (released || !loading || !sameLoadingUrl(url)) {
                return;
            }
            settleLoading();
        }

        @Override
        public void onReceivedError(WebView view, int errorCode, String description, String failingUrl) {
            if (!released && loading && sameLoadingUrl(failingUrl)) {
                settleLoading();
            }
        }

        @Override
        public void onReceivedError(WebView view, WebResourceRequest request, WebResourceError error) {
            if (!released && request != null && request.isForMainFrame() && loading
                    && sameLoadingUrl(request.getUrl().toString())) {
                settleLoading();
            }
        }

        @Override
        public void onReceivedHttpError(WebView view, WebResourceRequest request, android.webkit.WebResourceResponse errorResponse) {
            if (!released && request != null && request.isForMainFrame() && loading
                    && sameLoadingUrl(request.getUrl().toString())) {
                settleLoading();
            }
        }

        @Override
        public void onReceivedSslError(WebView view, SslErrorHandler handler, SslError error) {
            handleSslError(handler, error == null ? null : error.getUrl());
        }

        private void handleSslError(SslErrorHandler handler, String errorUrl) {
            if (handler != null) {
                handler.cancel();
            }
            // Always reject the certificate. Only an error for the active navigation may
            // settle loading; stale or malformed callbacks must not settle a newer load.
            if (!released && loading && errorUrl != null && sameLoadingUrl(errorUrl)) {
                settleLoading();
            }
        }

        private boolean sameLoadingUrl(String url) {
            return loadingUrl != null && loadingUrl.equals(url);
        }
    }

    /** Returns whether a main-frame URL is safe to keep inside this WebView. */
    static boolean isHttpOrHttpsUrl(String url) {
        if (url == null) {
            return false;
        }
        try {
            String scheme = Uri.parse(url).getScheme();
            return "http".equalsIgnoreCase(scheme) || "https".equalsIgnoreCase(scheme);
        } catch (RuntimeException ignored) {
            return false;
        }
    }

    /*
     * WebView does not attach a navigation generation to page-finished/error callbacks. A
     * same-URL callback from an older navigation is therefore indistinguishable from the live
     * navigation; URL matching is the strongest reliable guard available on API23+.
     */

    // Package-private callback seam for deterministic tests; production callbacks use the same path.
    void handleSslErrorForTest(SslErrorHandler handler, String errorUrl) {
        if (mWebViewClient != null) {
            mWebViewClient.handleSslError(handler, errorUrl);
        }
    }

    static final FrameLayout.LayoutParams COVER_SCREEN_PARAMS =
            new FrameLayout.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.MATCH_PARENT);
}

File: app/src/main/kotlin/com/goodtrendltd/HolySongs/HTML5WebView.kt
package com.goodtrendltd.HolySongs

import android.app.Activity
import android.app.ProgressDialog
import android.content.Context
import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.net.Uri
import android.net.http.SslError
import android.util.AttributeSet
import android.util.Log
import android.view.View
import android.view.ViewGroup
import android.view.Window
import android.webkit.GeolocationPermissions
import android.webkit.SslErrorHandler
import android.webkit.WebChromeClient
import android.webkit.WebResourceError
import android.webkit.WebResourceRequest
import android.webkit.WebResourceResponse
import android.webkit.WebSettings
import android.webkit.WebView
import android.webkit.WebViewClient
import android.widget.FrameLayout
import android.widget.LinearLayout
import android.widget.ProgressBar
import android.widget.TextView

class HTML5WebView : WebView {
    private var mContext: Context? = null
    private var mWebChromeClient: MyWebChromeClient? = null
    private var mWebViewClient: MyWebViewClient? = null
    private var mCustomView: View? = null
    private var mCustomViewContainer: FrameLayout? = null
    private var mCustomViewCallback: WebChromeClient.CustomViewCallback? = null

    private var mContentView: FrameLayout? = null
    private var mBrowserFrameLayout: FrameLayout? = null
    private var mLayout: FrameLayout? = null

    private var pd: ProgressDialog? = null
    private var loadingUrl: String? = null
    private var loading = false
    private var hostActive = true
    private var released = false
    private var destroyed = false
    private var handlingProgressCancel = false
    private var ignoreNextPageStart = false

    private fun init(context: Context) {
        mContext = context
        mLayout = FrameLayout(context)

        mBrowserFrameLayout = FrameLayout(context).apply {
            layoutParams = FrameLayout.LayoutParams(
                ViewGroup.LayoutParams.MATCH_PARENT,
                ViewGroup.LayoutParams.MATCH_PARENT,
            )
        }
        mContentView = FrameLayout(context).apply {
            id = R.id.main_content
            layoutParams = FrameLayout.LayoutParams(
                ViewGroup.LayoutParams.MATCH_PARENT,
                ViewGroup.LayoutParams.MATCH_PARENT,
            )
        }
        mCustomViewContainer = FrameLayout(context).apply {
            id = R.id.fullscreen_custom_content
            layoutParams = FrameLayout.LayoutParams(
                ViewGroup.LayoutParams.MATCH_PARENT,
                ViewGroup.LayoutParams.MATCH_PARENT,
            )
            visibility = View.GONE
        }

        val outer = LinearLayout(context).apply {
            orientation = LinearLayout.VERTICAL
            layoutParams = FrameLayout.LayoutParams(
                ViewGroup.LayoutParams.MATCH_PARENT,
                ViewGroup.LayoutParams.MATCH_PARENT,
            )
        }
        val errorConsole = LinearLayout(context).apply {
            id = R.id.error_console
            orientation = LinearLayout.HORIZONTAL
            layoutParams = LinearLayout.LayoutParams(
                ViewGroup.LayoutParams.MATCH_PARENT,
                ViewGroup.LayoutParams.WRAP_CONTENT,
            )
        }
        outer.addView(errorConsole)
        outer.addView(mContentView, FrameLayout.LayoutParams(
            ViewGroup.LayoutParams.MATCH_PARENT,
            ViewGroup.LayoutParams.MATCH_PARENT,
        ))

        mBrowserFrameLayout!!.addView(mCustomViewContainer, COVER_SCREEN_PARAMS)
        mBrowserFrameLayout!!.addView(outer, COVER_SCREEN_PARAMS)
        mLayout!!.addView(mBrowserFrameLayout, COVER_SCREEN_PARAMS)

        mWebChromeClient = MyWebChromeClient()
        webChromeClient = mWebChromeClient!!

        mWebViewClient = MyWebViewClient()
        webViewClient = mWebViewClient!!

        val settings: WebSettings = settings
        settings.layoutAlgorithm = WebSettings.LayoutAlgorithm.NARROW_COLUMNS
        settings.saveFormData = true
        settings.javaScriptEnabled = true
        settings.domStorageEnabled = true

        mContentView!!.addView(this)
    }

    constructor(context: Context) : super(context) {
        init(context)
    }

    constructor(context: Context, attrs: AttributeSet?) : super(context, attrs) {
        init(context)
    }

    constructor(context: Context, attrs: AttributeSet?, defStyle: Int) : super(context, attrs, defStyle) {
        init(context)
    }

    fun getLayout(): FrameLayout = mLayout ?: FrameLayout(context)

    fun inCustomView(): Boolean = mCustomView != null

    val isLoading: Boolean
        get() = loading

    internal val isReleased: Boolean
        get() = released

    internal fun getWebViewClientForTest(): WebViewClient = mWebViewClient!!

    internal fun getWebChromeClientForTest(): WebChromeClient = mWebChromeClient!!

    fun setHostActive(active: Boolean) {
        if (released) return
        hostActive = active
        if (!active) {
            dismissProgress()
        }
    }

    fun cancelLoading() {
        if (released) return
        ignoreNextPageStart = true
        stopLoading()
        settleLoading()
    }

    internal fun cancelProgressDialogForTest() {
        if (pd != null) {
            pd!!.cancel()
        } else {
            cancelLoading()
        }
    }

    private fun cancelLoadingFromProgressDialog() {
        if (released || handlingProgressCancel) return
        handlingProgressCancel = true
        try {
            cancelLoading()
        } finally {
            handlingProgressCancel = false
        }
    }

    fun pauseForHost() {
        if (!released) {
            onPause()
        }
    }

    fun resumeForHost() {
        if (!released) {
            onResume()
        }
    }

    fun hideCustomView() {
        if (!released) {
            mWebChromeClient?.onHideCustomView()
        }
    }

    internal fun release() {
        if (released) return
        released = true
        hostActive = false
        settleLoading()
        hideCustomViewForRelease()
        if (mContentView != null) {
            mContentView!!.removeView(this)
        }
        if (parent is ViewGroup) {
            (parent as ViewGroup).removeView(this)
        }
        if (!destroyed) {
            destroyed = true
            super.destroy()
        }
        mContext = null
        mContentView = null
        mBrowserFrameLayout = null
        mCustomViewContainer = null
        mLayout = null
    }

    private fun hideCustomViewForRelease() {
        val view = mCustomView ?: return
        view.visibility = View.GONE
        if (mCustomViewContainer != null) {
            mCustomViewContainer!!.removeView(view)
            mCustomViewContainer!!.visibility = View.GONE
        }
        val callback = mCustomViewCallback
        mCustomView = null
        mCustomViewCallback = null
        callback?.onCustomViewHidden()
    }

    private fun settleLoading() {
        loading = false
        loadingUrl = null
        dismissProgress()
    }

    private fun dismissProgress() {
        val dialog = pd
        pd = null
        if (dialog != null && dialog.isShowing) {
            try {
                dialog.dismiss()
            } catch (_: RuntimeException) {
                // The host window may already have gone away.
            }
        }
    }

    private fun isLiveForUi(): Boolean {
        val context = mContext as? Activity ?: return false
        return !released && hostActive && !context.isFinishing
    }

    internal fun handleSslErrorForTest(handler: SslErrorHandler?, errorUrl: String?) {
        mWebViewClient?.handleSslError(handler, errorUrl)
    }

    private inner class MyWebChromeClient : WebChromeClient() {
        private var mDefaultVideoPoster: Bitmap? = null
        private var mVideoProgressView: View? = null

        override fun onShowCustomView(view: View, callback: WebChromeClient.CustomViewCallback) {
            if (released || !hostActive || view == null) {
                callback?.onCustomViewHidden()
                return
            }
            this@HTML5WebView.visibility = View.GONE
            if (mCustomView != null) {
                callback?.onCustomViewHidden()
                return
            }
            mCustomViewContainer!!.addView(view)
            mCustomView = view
            mCustomViewCallback = callback
            mCustomViewContainer!!.visibility = View.VISIBLE
        }

        override fun onHideCustomView() {
            val view = mCustomView ?: return
            view.visibility = View.GONE
            if (mCustomViewContainer != null) {
                mCustomViewContainer!!.removeView(view)
                mCustomViewContainer!!.visibility = View.GONE
            }
            val callback = mCustomViewCallback
            mCustomView = null
            mCustomViewCallback = null
            callback?.onCustomViewHidden()
            if (!released) {
                this@HTML5WebView.visibility = View.VISIBLE
            }
        }

        override fun getDefaultVideoPoster(): Bitmap? {
            if (mDefaultVideoPoster == null && !released) {
                mDefaultVideoPoster = BitmapFactory.decodeResource(resources, R.drawable.default_video_poster)
            }
            return mDefaultVideoPoster
        }

        override fun getVideoLoadingProgressView(): View? {
            if (mVideoProgressView == null && !released) {
                val context = mContext ?: return null
                mVideoProgressView = LinearLayout(context).apply {
                    id = R.id.progress_indicator
                    orientation = LinearLayout.VERTICAL
                    layoutParams = FrameLayout.LayoutParams(
                        ViewGroup.LayoutParams.WRAP_CONTENT,
                        ViewGroup.LayoutParams.WRAP_CONTENT,
                    )
                    gravity = android.view.Gravity.CENTER
                    addView(ProgressBar(context, null, android.R.attr.progressBarStyleLarge).apply {
                        isIndeterminate = true
                        layoutParams = LinearLayout.LayoutParams(
                            ViewGroup.LayoutParams.WRAP_CONTENT,
                            ViewGroup.LayoutParams.WRAP_CONTENT,
                        ).apply {
                            gravity = android.view.Gravity.CENTER
                        }
                    })
                    addView(TextView(context).apply {
                        text = context.getString(R.string.loading_video)
                        textSize = 14f
                        setTextColor(android.graphics.Color.DKGRAY)
                        setPadding(0, 5, 0, 0)
                        gravity = android.view.Gravity.CENTER
                        layoutParams = LinearLayout.LayoutParams(
                            ViewGroup.LayoutParams.WRAP_CONTENT,
                            ViewGroup.LayoutParams.WRAP_CONTENT,
                        ).apply {
                            gravity = android.view.Gravity.CENTER
                        }
                    })
                }
            }
            return mVideoProgressView
        }

        override fun onReceivedTitle(view: WebView, title: String) {
            if (isLiveForUi()) {
                (mContext as Activity).title = title
            }
        }

        override fun onProgressChanged(view: WebView, newProgress: Int) {
            if (!released && loading && newProgress >= 100) {
                settleLoading()
            }
            if (isLiveForUi()) {
                (mContext as Activity).window.setFeatureInt(Window.FEATURE_PROGRESS, newProgress * 100)
            }
        }

        override fun onGeolocationPermissionsShowPrompt(origin: String, callback: GeolocationPermissions.Callback) {
            callback.invoke(origin, false, false)
        }
    }

    private inner class MyWebViewClient : WebViewClient() {
        override fun shouldOverrideUrlLoading(view: WebView, url: String): Boolean {
            if (!isHttpOrHttpsUrl(url)) {
                return true
            }
            if (!released) {
                Log.i(LOGTAG, "shouldOverrideUrlLoading: $url")
            }
            return false
        }

        override fun shouldOverrideUrlLoading(view: WebView, request: WebResourceRequest): Boolean {
            val url = request.url?.toString()
            if (!isHttpOrHttpsUrl(url)) {
                return true
            }
            if (!released) {
                Log.i(LOGTAG, "shouldOverrideUrlLoading: $url")
            }
            return false
        }

        override fun onPageStarted(view: WebView, url: String?, favicon: Bitmap?) {
            if (released) {
                return
            }
            if (ignoreNextPageStart) {
                ignoreNextPageStart = false
                return
            }
            loading = true
            loadingUrl = url
            if (isLiveForUi()) {
                dismissProgress()
                try {
                    val context = mContext ?: return
                    pd = ProgressDialog.show(context, "", context.getString(R.string.loading_video), true)
                    pd?.setCancelable(true)
                    pd?.setOnCancelListener { cancelLoadingFromProgressDialog() }
                } catch (_: RuntimeException) {
                    pd = null
                }
            }
        }

        override fun onPageFinished(view: WebView, url: String?) {
            if (!released && loading && sameLoadingUrl(url)) {
                settleLoading()
            }
        }

        override fun onReceivedError(view: WebView, errorCode: Int, description: String?, failingUrl: String?) {
            if (!released && loading && sameLoadingUrl(failingUrl)) {
                settleLoading()
            }
        }

        override fun onReceivedError(view: WebView, request: WebResourceRequest?, error: WebResourceError) {
            if (!released && request != null && request.isForMainFrame && loading && sameLoadingUrl(request.url.toString())) {
                settleLoading()
            }
        }

        override fun onReceivedHttpError(
            view: WebView,
            request: WebResourceRequest?,
            errorResponse: WebResourceResponse,
        ) {
            if (!released && request != null && request.isForMainFrame && loading && sameLoadingUrl(request.url.toString())) {
                settleLoading()
            }
        }

        override fun onReceivedSslError(view: WebView, handler: SslErrorHandler, error: SslError?) {
            handleSslError(handler, error?.url)
        }

        internal fun handleSslError(handler: SslErrorHandler?, errorUrl: String?) {
            handler?.cancel()
            if (!released && loading && errorUrl != null && sameLoadingUrl(errorUrl)) {
                settleLoading()
            }
        }

        private fun sameLoadingUrl(url: String?): Boolean {
            return loadingUrl != null && loadingUrl == url
        }
    }

    companion object {
        const val LOGTAG: String = "HTML5WebView"

        @JvmField
        val COVER_SCREEN_PARAMS: FrameLayout.LayoutParams = FrameLayout.LayoutParams(
            ViewGroup.LayoutParams.MATCH_PARENT,
            ViewGroup.LayoutParams.MATCH_PARENT,
        )

        @JvmStatic
        fun isHttpOrHttpsUrl(url: String?): Boolean {
            if (url == null) {
                return false
            }
            return try {
                val scheme = Uri.parse(url).scheme
                "http".equals(scheme, ignoreCase = true) || "https".equals(scheme, ignoreCase = true)
            } catch (_: RuntimeException) {
                false
            }
        }
    }
}
